pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -839,6 +820,95 @@ func (volTopology *nodeVolumeTopology) GetNodeTopologyLabels(ctx context.Context
+	nodeInfo *commoncotypes.NodeInfo,
+	csiNodeTopology *csinodetopologyv1alpha1.CSINodeTopology) *csinodetopologyv1alpha1.CSINodeTopology {
+	log := logger.GetLogger(ctx)
+	// If CSINodeTopology instance already exists, check if the NodeUUID
+	// parameter in Spec is populated. If not, patch the instance.
+	if csiNodeTopology.Spec.NodeUUID == "" ||
+		csiNodeTopology.Spec.NodeUUID != nodeInfo.NodeID {
+		if csiNodeTopology.Spec.NodeUUID == "" {
+			log.Infof("CSINodeTopology instance: %q with empty nodeUUID found. "+
+				"Patching the instance with nodeUUID", nodeInfo.NodeName)
+		} else {
+			log.Infof("CSINodeTopology instance: %q with different "+
+				"nodeUUID: %s found. Patching the instance with nodeUUID: %s",
+				nodeInfo.NodeName, csiNodeTopology.Spec.NodeUUID, nodeInfo.NodeID)
+		}
+		csiNodeTopology.Spec.NodeID = nodeInfo.NodeName
+		csiNodeTopology.Spec.NodeUUID = nodeInfo.NodeID
+	}
+	return csiNodeTopology
+}
+
+func (volTopology *nodeVolumeTopology) patchCSINodeTopology(
+	ctx context.Context,
+	oldTopo, newTopo *csinodetopologyv1alpha1.CSINodeTopology) (*csinodetopologyv1alpha1.CSINodeTopology, error) {
+	patch, err := getCSINodePatchData(oldTopo, newTopo, true)
+	if err != nil {
+		return oldTopo, err
+	}
+	rawPatch := client.RawPatch(types.MergePatchType, patch)
+	err = volTopology.csiNodeTopologyK8sClient.Patch(ctx, oldTopo, rawPatch)
+	if err != nil {
+		return oldTopo, err
+	}
+	return newTopo, nil
+}
+
+func getCSINodePatchData(
+	oldNodeTopology, newNodeTopology *csinodetopologyv1alpha1.CSINodeTopology,
+	addResourceVersionCheck bool) ([]byte, error) {
+	patchBytes, err := getPatchData(oldNodeTopology, newNodeTopology)
+	if err != nil {
+		return nil, err
+	}
+	if addResourceVersionCheck {
+		patchBytes, err = addResourceVersion(patchBytes, oldNodeTopology.ResourceVersion)
+		if err != nil {
+			return nil, fmt.Errorf("apply ResourceVersion to patch data failed: %v", err)
+		}
+	}
+	return patchBytes, nil
+}
+
+func addResourceVersion(patchBytes []byte, resourceVersion string) ([]byte, error) {
+	var patchMap map[string]interface{}
+	err := json.Unmarshal(patchBytes, &patchMap)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshalling patch with %v", err)
+	}
+	u := unstructured.Unstructured{Object: patchMap}
+	a, err := apiMeta.Accessor(&u)
+	if err != nil {
+		return nil, fmt.Errorf("error creating accessor with  %v", err)
+	}
+	a.SetResourceVersion(resourceVersion)
+	versionBytes, err := json.Marshal(patchMap)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling json patch with %v", err)
+	}
+	return versionBytes, nil
+}
+
+func getPatchData(oldObj, newObj interface{}) ([]byte, error) {
+	oldData, err := json.Marshal(oldObj)
+	if err != nil {
+		return nil, fmt.Errorf("marshal old object failed: %v", err)
+	}
+	newData, err := json.Marshal(newObj)
+	if err != nil {
+		return nil, fmt.Errorf("marshal new object failed: %v", err)
+	}
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
```

Comment:
can we patch without creating additional dependency on the repository - github.com/evanphx/json-patch/v5 

Example. Here is how we have patched some resources in the vSphere CSI Driver.

https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/ffd73326a9e77bd9616c246d801a6683f79cc117/pkg/syncer/storagepool/storageclasswatch.go#L253

Reply:
That dependency is already in this repo. See - https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/go.sum#L211

I am happy to modify this change, but IMO patching with `resourceVersion` etc is more safer way to patch an object. Also - it is safer to calculate diffs using existing tools, rather than by hand.